### ISSUE TITLE
spaceCost and timeCost compared to other algorithms

### ISSUE THREAD

**Issue by samuel-lucas6** [samuel-lucas6/draft-lucas-balloon-hashing#22]
`timeCost` is a misleading name because increasing `spaceCost` also increases the delay. Then `spaceCost` is the number of blocks when it would be more intuitive if the memory size (e.g., in bytes or KiB) was specified. This saves the user having to compute the number of blocks from their target memory size and the hash function output length. Because the weird thing at the moment is that you can have the same `spaceCost` with two different hash functions and you'll get a different memory size.

The problem is what to call these (e.g., `iterations` and `memorySize` perhaps) and how changing `spaceCost` affects the domain separation.

**Comment by cipriancraciun** [comment]
With regard to `timeCost` the original paper actually calls it everywhere as "rounds".

And, at least given the cryptography context, I think its more clearer to anyone what "rounds" means.  (Or even "iterations", but I think that the whole literature uses the word "round".)

On the other side, calling something "time" might be misleading, because as with `spaceCost`, it doesn't actually translate to a physical unit, but instead the actual "time" depends on the `spaceCost`, the efficiency of the chosen hash, and hardware performance of the machine running the code.

----

With regard to `spaceCost`, I personally think of it as the "width" of the process.

However, with regard to the proposal to be expressed in MiB (or anything directly translatable into memory), is perhaps problematic because:
* given that `timeCost` is "logical", if `spaceCost` is physical, there is a mismatch between the two;  (they should be perhaps both physical or both logical;)
* given that the `timeCost` and `spaceCost` directly impact the number of hashes computed, they are perhaps less about "memory" and more about "total hash iterations";

----

Or, form another perspective, there could be two functions standardized:
* a generic one, that takes the generic hash function as input, and the "logical" `rouds` / `width` arguments;
* a set of standard instantiations of the generic function, with concrete hashes (say SHA3-512, Blake3(-256), etc.), which should take `memoryCost` in terms of MiB (and deduce the `width` automatically) and `timeCost` which should be tweaked in accordance to the speed or complexity of the hash function, so that keeping `memoryCost` and `timeCost` constant, and varying the standard instantiation, the total resource consumption (on a "standard computer") should be "kind of the same";

----

And finally, how about a "radical" approach (which seems to be the current "best practice") of providing a few pre-defined levels (or purposes), so that users don't shoot themselves in the foot:
* "authentication";  tailored for online authentication, where the total time should be under 100ms on an "average modile";  (or whatever the current UX standards are today;)
* "throttling";  for example when the hard-hash-function is used as a work-factor to thwart attackers and DoS;  (the focus is not on cryptographic strength, but on wasting the attacker's time and resources;)
* "NIST-level-1";  as per <https://csrc.nist.gov/projects/post-quantum-cryptography/post-quantum-cryptography-standardization/evaluation-criteria/security-(evaluation-criteria)> -- the amount of work needed to brute-force a password at this level, should "require computational resources comparable to or greater than those required for key search on a block cipher with a 128-bit key (e.g. AES128)>";
* "NIST-level-x" (2 to 5) -- similar as above;  (perhaps only level 3 and 5 which are defined in terms of AES, the others being defined in terms of SHA;)

**Comment by samuel-lucas6** [comment]
> With regard to timeCost the original paper actually calls it everywhere as "rounds".

The pseudocode and reference implementation use `t_cost` for time cost though.

> And, at least given the cryptography context, I think its more clearer to anyone what "rounds" means. (Or even "iterations", but I think that the whole literature uses the word "round".)

I like iterations because it matches PBKDF2. Argon2 calls it passes.

> On the other side, calling something "time" might be misleading, because as with spaceCost, it doesn't actually translate to a physical unit, but instead the actual "time" depends on the spaceCost, the efficiency of the chosen hash, and hardware performance of the machine running the code.

I definitely don't like `timeCost` and `spaceCost`.

> However, with regard to the proposal to be expressed in MiB (or anything directly translatable into memory), is perhaps problematic because:

The issue I realised with having KiB or MiB for the memory is that it doesn't work well with the power of 2 to avoid modulo bias, so I might have to leave the memory parameter as blocks.

> And finally, how about a "radical" approach (which seems to be the current "best practice") of providing a few pre-defined levels (or purposes), so that users don't shoot themselves in the foot

This isn't a good idea because of the performance difference depending on the implementation/device and how the parameters should be adjusted over time.

**Comment by jedisct1** [comment]
> The issue I realised with having KiB or MiB for the memory is that it doesn't work well with the power of 2 to avoid modulo bias, so I might have to leave the memory parameter as blocks.

The problem here is that applications don't have a straightforward way to predict how much memory is going to be required to run the algorithm. The block size should be an internal implementation/algorithm detail, not something users have to understand to use BKDF correctly.

**Comment by samuel-lucas6** [comment]
> The problem here is that applications don't have a straightforward way to predict how much memory is going to be required to run the algorithm. The block size should be an internal implementation/algorithm detail, not something users have to understand to use BKDF correctly.

Yeah, I agree this should be changed. It's just how best to do it.
